ddht/v5_1/session.py

```diff
@@ -159,6 +159,15 @@ def keys(self) -> SessionKeys:
     async def _process_message_buffers(self) -> None:
         ...
 
+    async def await_handshake_completion(self) -> None:
+        if self.is_after_handshake:
+            raise Exception("Handshake is already complete")
+
+        async with self._events.session_handshake_complete.subscribe() as subscription:
```

Comment:
Rather than this, you should use a `trio.Event()`.  This method would simply be:

```python
def __init__(self, ...):
    self._handshake_complete = trio.Event()
async def await_handshake_completion(self) -> None:
    await self._handshake_complete.wait()
```

Then, in the two places where we do `self._status = SessionStatus.AFTER` we can also do `self._handshake_complete.set()`.